Allow NodeTimer and ABM interval configuration.

sofar opened the pull request:
ABM's are hardcoded to run every 2.0s, NodeTimers are hard coded
to run at every 1.0s.

However, these timers can be better tuned for both higher and
lower values by server owners. Some server owners want to, and
have the resources to send more packets per second to clients,
and so they may wish to send smaller updates sooner. Right now
all ABM's are coalesced into 2.0 second intervals, resulting
in large send queues to all clients. By reducing the amount
of possible timers, one can get a far better response rate
and lower the perception of lag.

On the other side of the camp, some servers may want to increase
these values, which again isn't easily doable.

The global settings abm_interval and nodetimer_interval are
set to current values by default. I've tested with 0.2/0.5
type values and noticed a greatly improved response and better
scattering of nodetimers, as well as enjoying not faceplanting
into doors with pressure plates anymore.

0-afflatus commented:
So I have to ask then, what does the 'interval' setting do in minetest.register_abm?

rubenwardy commented on src/environment.cpp:
```diff
@@ -1064,7 +1064,8 @@ void ServerEnvironment::step(float dtime)
 		Manage active block list
 	*/
-	if(m_active_blocks_management_interval.step(dtime, 2.0))
+	float abm_interval = g_settings->getFloat("abm_interval");
```
This should probably be cached using `static`

sofar replied:
That would make it unchangeable at runtime.

I asked specifically on IRC and got this reply from @celeron55:

> sofar: i would say calling it exactly once per setting for every step is fine, and even recommended so that the settings can be updated at runtime

Notice I'm properly reusing the setting in the rest of the function, so, I think this is appropriate. (I referenced your "cache a few settings" patch for this, as well, so I'm aware that eliminating settings-gets is also desirable).

nerzhul replied:
setting getter has a bad performance. Caching is required, especially on ServerEnv::step.
Also modify the abm_interval at runtime can be dangerous

sofar replied:
@nerzhul I don't see why this would be dangerous, modification would come through a chatcommand, and not apply until the next ```::step```, so nothing unsafe about that.

The same function already uses settings->get several times on these:
- dedicated_server_step
- active_block_range
and those aren't cached either.

Given @celeron55's comment, you'll have to make up your mind together, I couldn't really care either way (uncached is fine, but then there should be no uncached settings there either), but I can't change the code BOTH ways.

sofar commented:
The server doesn't process expired ABM's every server step. It by default only checks for expired ABM's every 2.0 seconds. The ABM's therefore are executed up to 2 seconds later than when they expired.

paramat commented:
Interesting, would it then be a good idea to change all 1-second interval ABMs to 2-second?

sofar commented:
No, I actually think that the long intervals are *hurtful* and these values should be much, much lower to begin with.

The problem is pretty easy to see. Suppose we have 2 ABM's running with a 2 second interval. With the current settings, they will always run exactly on the same time. This means that there was an opportunity to run 1 abm, send packets to clients, sleep a little, run the second abm, send packets to clients, it's now always going to run both abms, and send much bigger packets to clients.

In a server that's not under any stress, you shouldn't coalesce events - better to do small amounts of work and be done faster. So, it's better to attempt to run the ABM's out-of-sync.

If the server *is* under heavy stress, then now you have to deal with MORE work, and send LARGER updates to all clients. So, it's better to attempt to split up the work and run the ABM's out-of-sync.

As for NodeTimers, these are deliberately (just like ABM's) coded as ```float time```, and some mods even use fractional time here, expecting them to work faster. While potentially causing more NodeTimer events, I see no reason to *prevent* folks from trying this.

So ultimately, I chose to make these values *configurable* first. Once they are configurable, server ops have the option of making their game more responsive... or doing the opposite.

I thus expect that ABM/NodeTimer handling should be done way more often, so I strongly suggest we don't lower ABM rates, that would only make the coalescing problem worse.

paramat commented:
> ABM's are hardcoded to run every 5.0s
Did you mean 2.0s here?

sofar commented:
@paramat right, yes.

ShadowNinja commented on builtin/settingtypes.txt:
```diff
@@ -781,6 +781,12 @@ sqlite_synchronous (Synchronous SQLite) enum 2 0,1,2
+
+#    Length of time between NodeTimer execution cycles
+nodetimer_interval (NodeTimer interval) float 1.0
+
```
minetest.conf.example and the gettext dummy file should be updated too.

sofar replied:
Why isn't minetest.conf.example autogenerated somehow?

sofar replied:
oh, IC. nvm.

sofar commented:
@ShadowNinja updated conf example and gettext dummy, as well.

sofar commented:
@rubenwardy @nerzhul the settings are now cached. I'd rather see this merged faster without caching, then not.

est31 commented on src/environment.cpp:
```diff
@@ -1322,7 +1324,7 @@ void ServerEnvironment::step(float dtime)
 	*/
-	if(m_active_blocks_management_interval.step(dtime, 2.0))
+	if (m_active_blocks_management_interval.step(dtime, m_cache_abm_interval))
 	{
```
brace on same line

est31 commented on src/environment.cpp:
```diff
@@ -1399,11 +1401,11 @@ void ServerEnvironment::step(float dtime)
 	*/
-	if(m_active_blocks_nodemetadata_interval.step(dtime, 1.0))
+	if (m_active_blocks_nodemetadata_interval.step(dtime, m_cache_nodetimer_interval))
 	{
```
here too

est31 commented on src/environment.cpp:
```diff
@@ -1446,8 +1448,7 @@ void ServerEnvironment::step(float dtime)
 
-	const float abm_interval = 1.0;
-	if(m_active_block_modifier_interval.step(dtime, abm_interval))
+	if(m_active_block_modifier_interval.step(dtime, m_cache_abm_interval))
```
Space before (

est31 commented on src/environment.cpp:
```diff
@@ -1322,7 +1324,7 @@ void ServerEnvironment::step(float dtime)
-	if(m_active_blocks_management_interval.step(dtime, 2.0))
+	if (m_active_blocks_management_interval.step(dtime, m_cache_abm_interval))
 	{
 		ScopeProfiler sp(g_profiler, "SEnv: manage act. block list avg /2s", SPT_AVG);
```
Replace the 2s here with a dynamic string

sofar replied:
that means allocating a std::string every serverstep. Do you really want to do that? How about a generic string such as ```SEnv: manage act. block list avg /int``` instead?

est31 replied:
Okay, what about "interval" instead of "int"

sofar replied:
ok, I'll do that.

est31 commented on src/environment.cpp:
```diff
@@ -1399,11 +1401,11 @@ void ServerEnvironment::step(float dtime)
-	if(m_active_blocks_nodemetadata_interval.step(dtime, 1.0))
+	if (m_active_blocks_nodemetadata_interval.step(dtime, m_cache_nodetimer_interval))
 	{
 		ScopeProfiler sp(g_profiler, "SEnv: mess in act. blocks avg /1s", SPT_AVG);
```
dynamic string needed here too

est31 commented on src/environment.cpp:
```diff
@@ -1458,7 +1459,7 @@ void ServerEnvironment::step(float dtime)
 		TimeTaker timer("modify in active blocks");
```
dynamic string required in the line above this one.

est31 commented:
Other than that seems fine; :+1:  if fixed

sofar commented:
Updated with style and profiler string changes.

sofar commented:
@shadowninja, care to review?

ShadowNinja commented on builtin/settingtypes.txt:
```diff
@@ -781,6 +781,12 @@ sqlite_synchronous (Synchronous SQLite) enum 2 0,1,2
 dedicated_server_step (Dedicated server step) float 0.1
 
+#    Length of time between ABM execution cycles
+abm_interval (ABM modifier interval) float 2.0
```
"ABM modifier" == "Active Block Modifier modifier"

ShadowNinja commented on src/environment.cpp:
```diff
@@ -1322,9 +1324,8 @@ void ServerEnvironment::step(float dtime)
-	{
-		ScopeProfiler sp(g_profiler, "SEnv: manage act. block list avg /2s", SPT_AVG);
+	if (m_active_blocks_management_interval.step(dtime, m_cache_abm_interval)) {
+		ScopeProfiler sp(g_profiler, "SEnv: manage act. block list avg /interval", SPT_AVG);
```
`/2s` makes sense, but `/interval` doesn't make much sense since it isn't a number.  Just use `per interval`.

More importantly, this doesn't seem to actually be the ABM interval, but the management interval.  The ABM interval (below) was 1s but this was 2s.  Therefore, it seems like this will actually make ABMs coalesce *more*, by changing the ABM interval from 1s to 2s to match the management interval.  Is this correct?

sofar replied:
hmm, right, my idea is to experiment with far lower values (or eliminate the intervals entirely).

Would you agree to split this out into block_management_interval & abm_interval, then?

sofar replied:
names are just a suggestion, of course. could be active_block_mgmt_interval as well, or similar

ShadowNinja replied:
Yes, that would be fine.

ShadowNinja commented on src/environment.cpp:
```diff
@@ -1399,11 +1400,10 @@ void ServerEnvironment::step(float dtime)
-	{
-		ScopeProfiler sp(g_profiler, "SEnv: mess in act. blocks avg /1s", SPT_AVG);
+	if (m_active_blocks_nodemetadata_interval.step(dtime, m_cache_nodetimer_interval)) {
+		ScopeProfiler sp(g_profiler, "SEnv: mess in act. blocks avg /interval", SPT_AVG);
```
`/interval`

ShadowNinja commented on src/environment.cpp:
```diff
@@ -1446,19 +1446,18 @@ void ServerEnvironment::step(float dtime)
-		ScopeProfiler sp(g_profiler, "SEnv: modify in blocks avg /1s", SPT_AVG);
-		TimeTaker timer("modify in active blocks");
+		ScopeProfiler sp(g_profiler, "SEnv: modify in blocks avg /interval", SPT_AVG);
+		TimeTaker timer("modify in active blocks /interval");
```
`/interval`

ShadowNinja commented:
@sofar: Done.  I noticed what seemed like a substantial code issue and two language issues.

sofar commented:
- Added active_block_mgmt_interval (defaults to 2.0 secs)
- Modified abm_interval to default to 1.0 secs
- Fixed using the right value in environment.cpp
- Updated templates, docs

and last

- Changed scope to say ```... per``` and not ```... per interval```, since that was a lot of characters for very little information. I think people can understand that it's per (unit) and that the unit is (interval), and it won't make the scope text overflow to the right.

paramat commented:
Good feature i would like this.

celeron55 commented:
@sofar Why did you change "/" to "per" to begin with; I think "/int" would work well in place of /2s or /1s. "per" is very weird.

rubenwardy commented:
`per` doesn't sound right, but `per int`, `per inte` or `per interval` does.

sofar commented:
@celeron55 ```per interval``` made entries' text overflow the right margin.

how about ```(per)``` ?

Anyway, I'm not attached to how this is worded - I can live with ```per interval``` if people don't mind the overflow.

sofar commented:
Changed it back to ```per interval```. It does overflow the column margin again. Please review so this can merge.

paramat commented:
:+1:

paramat commented:
c5c727d62735e334ec64f07c557a2896d6d66d9f
